### PR TITLE
Transform Polls to Documents

### LINKED ISSUE #311: Polls should become documents
Polls should become a subclass of documents. This enables using the markdown editor, including adding links, images, ...

### PR DESCRIPTION
**Pull request by Bartzi**
Polls are now Documents (yeah!)

fixes #311 and #282

### CODE REVIEW

File: _1327/documents/models.py [hunk 1]
```diff
@@ -58,6 +58,9 @@ def get_view_permission(klass):
 		permission = "{}.{}".format(app_label, klass.VIEW_PERMISSION_NAME)
 		return permission
 
+	def save_formset(self, formset):
+		pass
+
```
Comment (Nef10) [inline _1327/documents/models.py]: 👍

File: _1327/documents/utils.py [hunk 2]
```diff
@@ -53,13 +53,7 @@ def handle_edit(request, document, formset=None, initial=None):
...
 						document.groups.add(group)
 				document.save()
-				if formset:
-					guests = formset.save(commit=False)
-					for guest in formset.deleted_objects:
-						guest.delete()
-					for guest in guests:
-						guest.minute = document
-						guest.save()
+				document.save_formset(formset)
```
Comment (Nef10) [inline _1327/documents/utils.py]: 👍

File: _1327/minutes/models.py [hunk 3]
```diff
@@ -106,6 +106,14 @@ def meta_information_html(self):
...
 		return template.render(Context({'document': self}))
 
+	def save_formset(self, formset):
+		guests = formset.save(commit=False)
+		for guest in formset.deleted_objects:
+			guest.delete()
+		for guest in guests:
+			guest.minute = self
+			guest.save()
+
```
Comment (Nef10) [inline _1327/minutes/models.py]: 👍

File: _1327/polls/models.py [hunk 4]
```diff
@@ -1,24 +1,35 @@
...
-	start_date = models.DateField()
-	end_date = models.DateField()
+class Poll(Document):
+
+	def can_be_changed_by(self, user):
+		permission_name = 'change_poll'
+		return user.has_perm(permission_name, self) or user.has_perm(permission_name)
+
+	start_date = models.DateField(default=datetime.now, verbose_name=_("Start Date"))
+	end_date = models.DateField(default=datetime.now, verbose_name=_("End Date"))
```
Comment (Nef10) [inline _1327/polls/models.py]: 👍

File: _1327/templates/base.html [hunk 5]
```diff
@@ -102,10 +102,10 @@
 									<li><a href="{% url 'documents:create' 'informationdocument' %}">{% trans "Information page" %}</a></li>
 								{% endif %}
 								{% if CAN_CREATE_MINUTES %}
-									<li><a href="{% url 'documents:create' 'myinutesdocument' %}">{% trans "Minutes" %}</a></li>
+									<li><a href="{% url 'documents:create' 'minutesdocument' %}">{% trans "Minutes" %}</a></li>
```
Comment (Nef10) [inline _1327/templates/base.html]: good catch

File: _1327/polls/urls.py [hunk 6]
```diff
@@ -3,8 +3,6 @@
 
 urlpatterns = patterns('_1327.polls.views',  # noqa
 	url(r"^$", "list", name="list"),
-	url(r"create$", "create", name="create"),
-	url(r"(?P<poll_id>\d+)/edit", "edit", name="edit"),
-	url(r"(?P<poll_id>\d+)/results$", "results", name="results"),
-	url(r"(?P<poll_id>\d+)/vote$", "vote", name="vote"),
+	url(r"(?P<url_title>[\w-]+)/results$", "results", name="results"),
+	url(r"(?P<url_title>[\w-]+)/vote$", "vote", name="vote"),
```
Comment (Nef10) [inline _1327/polls/urls.py]: 👍

File: _1327/polls/templates/polls_index.html [hunk 7]
```diff
@@ -89,12 +89,12 @@
...
 			{% for poll in finished_polls %}
 				<tr>
-					<td><a href="{% url "polls:results" poll.id %}">{{ poll.title }}</a></td>
+					<td><a href="{% url "polls:results" poll.url_title %}">{{ poll.title }}</a></td>
 					<td>{{ poll.start_date }} - {{ poll.end_date }}</td>
 					<td class="pull-right">
 						{% get_obj_perms request.user for poll as "poll_perms" %}
 						{% if "change_poll" in poll_perms %}
-							<a class="btn btn-warning btn-xs" href="{% url "polls:edit" poll.id %}">{% trans "Edit Poll" %}</a>
+							<a class="btn btn-warning btn-xs" href="{% url "documents:edit" poll.url_title %}">{% trans "Edit Poll" %}</a>
```
Comment (Nef10) [inline _1327/polls/templates/polls_index.html]: 👍

File: _1327/polls/forms.py [hunk 8]
```diff
@@ -15,5 +23,5 @@ class Meta:
 		exclude = ["poll", "votes"]
 		widgets = {
 			"description": forms.TextInput(),
-			"index": forms.HiddenInput()
+			# "index": forms.HiddenInput()
```
Comment (Nef10) [inline _1327/polls/forms.py]: ?

**Comment by Nef10** [comment]
migration does not work for me:
```
django.db.utils.IntegrityError: insert or update on table "polls_poll" violates foreign key constraint "polls_document_ptr_id_734a6c9a017bea32_fk_documents_document_id"
DETAIL:  Key (document_ptr_id)=(0) is not present in table "documents_document".
```

File: _1327/documents/utils.py [hunk 9]
```diff
@@ -155,3 +149,9 @@ def permission_warning(user, content_type, document):
 	anonymous_rights = get_anonymous_user().has_perm(content_type.model_class().VIEW_PERMISSION_NAME, document)
 	edit_rights = user.has_perm("change_informationdocument", document)
 	return edit_rights and not anonymous_rights
+
+
+def get_model_function(content_type, function_name):
+	# TODO add caching strategy?
```
Comment (Nef10) [inline _1327/documents/utils.py]: open a ticket for this

File: _1327/polls/tests.py [hunk 10]
```diff
@@ -292,20 +330,20 @@ def test_start_vote_single_choice_poll(self):
...
+		response = self.app.get(reverse('polls:vote', args=[self.poll.url_title]), user=self.user)
 		self.assertEqual(response.status_code, 200)
 		for choice in self.poll.choices.all():
 			self.assertIn(choice.text.encode('utf-8'), response.body)
 
 	def test_vote_without_submitting_a_choice(self):
-		response = self.app.post(reverse('polls:vote', args=[self.poll.id]), user=self.user)
-		self.assertRedirects(response, reverse('polls:vote', args=[self.poll.id]))
+		response = self.app.post(reverse('documents:view', args=[self.poll.url_title]), user=self.user)
+		self.assertRedirects(response, reverse('documents:view', args=[self.poll.url_title]))
```
Comment (Nef10) [inline _1327/polls/tests.py]: ?

File: _1327/polls/tests.py [hunk 11]
```diff
@@ -315,8 +353,8 @@ def test_vote_single_choice_submitting_more_than_one_choice(self):
 		for choice in self.poll.choices.all():
 			data.append(('choice', choice.id))
 
-		response = self.app.post(reverse('polls:vote', args=[self.poll.id]), params=data, user=self.user)
-		self.assertRedirects(response, reverse('polls:vote', args=[self.poll.id]))
+		response = self.app.post(reverse('documents:view', args=[self.poll.url_title]), params=data, user=self.user)
+		self.assertRedirects(response, reverse('documents:view', args=[self.poll.url_title]))
```
Comment (Nef10) [inline _1327/polls/tests.py]: ?

File: _1327/polls/tests.py [hunk 12]
```diff
@@ -341,17 +379,17 @@ def test_vote_multiple_choice_correctly(self):
...
-		self.assertRedirects(response, reverse('polls:results', args=[self.poll.id]))
+		self.assertRedirects(response, reverse('polls:results', args=[self.poll.url_title]))
 
 	def test_view_before_poll_has_started(self):
 		self.poll.start_date += datetime.timedelta(weeks=1)
 		self.poll.save()
 
-		response = self.app.get(reverse('polls:vote', args=[self.poll.id]), user=self.user, expect_errors=True)
+		response = self.app.get(reverse('polls:vote', args=[self.poll.url_title]), user=self.user, expect_errors=True)
 		self.assertEqual(response.status_code, 404)
```
Comment (Nef10) [inline _1327/polls/tests.py]: two time you changed `polls:vote` into `documents:view`, why?
otherwise :+1:

Comment (Bartzi) [inline _1327/polls/tests.py]: yeah right! I actually think that the `polls:vote` and `polls:view` views should not be there anymore I'd like to refactor this when this is done. For now everything works fine, but having just one view that decides on its own which view to show to the user is better I think.

**Comment by Bartzi** [comment]
I think the migration does not work because of the transition of Polls to Documents. The database is changed such that every poll now points to a document and if you have any polls in your database the system will not create any new Documents for you while migrating. I think the best solution is to simply delete the database and start over again. Luckily we are still able to do this at this point ;)

File: _1327/documents/utils.py (hunk 9 above)
Comment (Bartzi) [inline _1327/documents/utils.py]: I've created a new ticket for this. See #328

File: _1327/polls/models.py [hunk 13]
```diff
@@ -35,6 +46,31 @@ def url(self, id):
...
+	def save_formset(self, formset):
+		choices = formset.save(commit=False)
+		for choice in formset.deleted_objects:
+			choice.delete()
+		for choice in choices:
+			choice.poll = self
+			choice.save()
+
+	@property
+	def votes(self):
```
Comment (janno42) [inline _1327/polls/models.py]: `num_votes`

**Comment by janno42** [comment]
please only allow positive numbers for `Max allowed number of answers` ;)

File: _1327/polls/models.py [hunk 14]
```diff
@@ -35,6 +46,31 @@ def url(self, id):
...
+		return self.choices.all().aggregate(Sum('votes')).popitem()[1]
```
Comment (janno42) [inline _1327/polls/models.py]: `self.choices.aggregate(Sum('votes')).get('votes__sum')`

**Comment by janno42** [comment]
once the tests succeed: :package:

**Pull request merged**